### PR TITLE
[now-cli] Add warning for DNS record if domain exists in subdomain

### PR DESCRIPTION
**Pull request by paulogdm**
Add warning for DNS record if domain exists in subdomain

### CODE REVIEW

File: packages/now-cli/src/commands/dns/add.ts [hunk 1]
```diff
@@ -54,9 +56,23 @@ export default async function add(
     );
     return 1;
   }
+ 
+  const { domain, data: argData } = parsedParams;
+  if (argData && domain === argData.name.substr(argData.name.length - domain.length, argData.name.length)) {
```
Comment (TooTallNate) [inline packages/now-cli/src/commands/dns/add.ts]: Is this the same logic?

```suggestion
  if (argData && argData.name.endsWith(domain)) {
```

File: packages/now-cli/src/commands/dns/add.ts [hunk 2]
```diff
@@ -117,3 +133,63 @@ export default async function add(
 
   return 0;
 }
+
+function readConfirmation(
```
Comment (TooTallNate) [inline packages/now-cli/src/commands/dns/add.ts]: We already have a function for this. See `src/util/prompt-bool.ts`.

File: packages/now-cli/src/commands/dns/add.ts [hunk 3]
```diff
@@ -117,3 +133,63 @@ export default async function add(
...
+            .toString()
+            .trim()
+            .toLowerCase() === 'y'
+        );
+      })
+      .resume();
+  });
+}
+
+function getAddebleTableRow(domain: string, argData: DNSRecordData) {
```
Comment (TooTallNate) [inline packages/now-cli/src/commands/dns/add.ts]: I'm not sure what "Addeble" means. Is that a typo?

**Comment by paulogdm** [comment]
@TooTallNate Still WIP. Sorry for the early review. I will change a few things right away.

File: packages/now-cli/src/commands/dns/add.ts (hunk 3 above)
Comment (paulogdm) [reply packages/now-cli/src/commands/dns/add.ts]: It is.

File: packages/now-cli/src/commands/dns/add.ts (hunk 2 above)
Comment (paulogdm) [reply packages/now-cli/src/commands/dns/add.ts]: I will check that. Thanks!

File: packages/now-cli/src/commands/dns/add.ts (hunk 1 above)
Comment (paulogdm) [reply packages/now-cli/src/commands/dns/add.ts]: Nice catch!

**Comment by paulogdm** [comment]
Ok, I did all of this because I couldn't resolve the following error on cb65b587a740efb500f3b7b76784513f64d188b1:
```
Error: [tsl] ERROR in /home/circleci/repo/packages/now-cli/src/commands/dns/add.ts(69,49)
      TS2339: Property 'value' does not exist on type 'DNSRecordData'.
  Property 'value' does not exist on type 'SRVRecordData'.
```
After talking with @styfle , it was clear this is due to the following structure:
```
export type DNSRecordData =
  {
      name: string;
      type: string;
      value: string;
    } ...
```
So I enumareted types for each record to avoid a generic `type: string`.

**tl;dr** https://media0.giphy.com/media/llbEoVMhkLngWlzVVa/giphy.gif

File: packages/now-cli/src/util/dns/get-dns-data.ts [hunk 4]
```diff
@@ -25,64 +35,86 @@ export default async function getDNSData(
...
+      case 'AAAA':
+        return (await getData(output, type)) as AAAARecordData
+      case 'ALIAS':
+        return (await getData(output, type)) as ALIASRecordData
+      case 'CAA':
+        return (await getData(output, type)) as CAARecordData
+      case 'CNAME':
+        return (await getData(output, type)) as CNAMERecordData
+      case 'TXT':
+        return (await getData(output, type)) as TXTRecordData
```
Comment (styfle) [inline packages/now-cli/src/util/dns/get-dns-data.ts]: Is it necessary to type cast `as` for all of these? The return type is still `DNSRecordData` so I think it is not necessary. In fact, you shouldn't need a `switch` at all here.

File: packages/now-cli/src/types.ts [hunk 5]
```diff
@@ -194,7 +194,7 @@ export type DNSRecord = {
   updated: number;
 };
 
-type SRVRecordData = {
+export type SRVRecordData = {
```
Comment (styfle) [inline packages/now-cli/src/types.ts]:
```suggestion
export interface SRVRecordData extends GenericDNSRecordData {
```

File: packages/now-cli/src/types.ts [hunk 6]
```diff
@@ -205,19 +205,49 @@ type SRVRecordData = {
   };
 };
 
-type MXRecordData = {
+export type MXRecordData = {
```
Comment (styfle) [inline packages/now-cli/src/types.ts]:
```suggestion
export interface MXRecordData extends GenericDNSRecordData {
```

File: packages/now-cli/src/types.ts [hunk 7]
```diff
@@ -194,7 +194,7 @@ export type DNSRecord = {
   updated: number;
 };
```
Comment (styfle) [inline packages/now-cli/src/types.ts:190]: Is the type `DNSRecord` used? Perhaps it should be deleted.

File: packages/now-cli/src/util/dns/get-dns-data.ts [hunk 8]
```diff
@@ -25,64 +35,86 @@ export default async function getDNSData(
...
-          `${mxPriority}`
-        )} ${chalk.cyan(value)}`
-      );
-      return (await verifyData())
-        ? {
-            name,
-            type,
-            value,
-            mxPriority
+async function getData(output: Output, type: string) {
```
Comment (styfle) [inline packages/now-cli/src/util/dns/get-dns-data.ts]:
```suggestion
async function getData(output: Output, type: string): Promise<DNSRecordData | null> {
```

File: packages/now-cli/src/util/dns/parse-add-dns-record-args.ts [hunk 9]
```diff
@@ -23,39 +33,68 @@ export default function parseAddArgs(
...
+            };
+          case 'CNAME':
+            return {
+              domain,
+              data: { name, value, type } as CNAMERecordData
+            };
+          case 'TXT':
+            return {
+              domain,
+              data: { name, value, type } as TXTRecordData
```
Comment (styfle) [inline packages/now-cli/src/util/dns/parse-add-dns-record-args.ts]: Is `as` necessary here?

File: packages/now-cli/src/util/dns/get-dns-data.ts (hunk 4 above)
Comment (paulogdm) [reply packages/now-cli/src/util/dns/get-dns-data.ts]: I agree but I remove the cast TS complains about assignable types. A quick test is:
```
      case 'SRV':
        return (await getData(output, type)) as SRVRecordData
```
vs
```
      case 'SRV':
        return (await getData(output, type)) // erroring
```

File: packages/now-cli/src/types.ts (hunk 7 above)
Comment (paulogdm) [reply packages/now-cli/src/types.ts:190]: It is. I believe it is used on API responses.
`DNSRecord` => Receiving data from our API
`DNSRecordData` => Parsed record.

Since this is legacy, should we take a different approach to this as well?

Comment (styfle) [reply packages/now-cli/src/types.ts:190]: I'm not sure about legacy. I'll add @mglagola as a reviewer who should know how DNS APIs work.

**Comment by paulogdm** [comment]
One alternative to get this "warning" through without major type changes would be to simply have this piece of code:
```
 
  const { domain, data: argData } = parsedParams;
  if (argData && argData.name.endsWith(domain)) {
    output.log(`'Domain identified inside the subdomain argument. The following record will be created':`);
    const anyArgData:any = argData;
    switch (anyArgData.type) {
      case 'MX':
        output.log(
          chalk.bold(`${anyArgData.type} ${anyArgData.value} ${anyArgData.mxPriority}`)
        );
        break;
      case 'SRV':
        output.log(
          chalk.bold(`${anyArgData.type} ${anyArgData.srv.priority} ${anyArgData.srv.port} ${anyArgData.srv.weight} ${anyArgData.srv.target}`)
        );
        break;
      default:
        output.log(
          chalk.bold(`${anyArgData.type} ${anyArgData.value}`)
        );
        break;
    }
    if (!await promptBool(`Do you want to create this record on ${anyArgData.name}.${domain}?`)) {
      output.log(`Aborted`);
      return 1;
    }
  }
```
avoiding the whole `type` unknown problem.

**Comment by styfle** [comment]
@paulogdm I pushed a change.

The issue is that `SRVRecordData` expects a `value` property but you did not assign it.

image

**Comment by paulogdm** [comment]
```
export interface SRVRecordData extends GenericDNSRecordData {
  name: string;
  type: 'SRV';
  srv: {
    port: number;
    priority: number;
    target: string;
    weight: number;
  };
};

export interface MXRecordData extends GenericDNSRecordData {
  name: string;
  type: 'MX';
  value: string;
  mxPriority: number;
};
```
I don't think they should extend `GenericDNSRecordData`, since they have different properties. Maybe `MXRecordData` can but `SRVRecordData` should not:
```
export interface SRVRecordData {
  name: string;
  type: 'SRV';
  srv: {
    port: number;
    priority: number;
    target: string;
    weight: number;
  };
};

export interface MXRecordData extends GenericDNSRecordData {
  mxPriority: number;
};
```

**Comment by styfle** [comment]
@paulogdm Sounds good to me 👍

**Comment by paulogdm** [comment]
Nice, just pushed ^

**Comment by paulogdm** [comment]
Going to check why it failed soon...

File: packages/now-cli/src/types.ts [hunk 10]
```diff
@@ -205,19 +205,46 @@ type SRVRecordData = {
   };
 };
 
-type MXRecordData = {
-  name: string;
-  type: 'MX';
-  value: string;
+export interface MXRecordData extends GenericDNSRecordData {
   mxPriority: number;
```
Comment (styfle) [inline packages/now-cli/src/types.ts:204]: You need to have `type: 'MX';` here

File: packages/now-cli/src/util/dns/parse-add-dns-record-args.ts (hunk 9 above)
Comment (styfle) [reply packages/now-cli/src/util/dns/parse-add-dns-record-args.ts]: Instead, you can do the following:

```ts
const data: TXTRecordData = { name, value, type };
return { domain, data };
```

File: packages/now-cli/src/types.ts (hunk 10 above)
Comment (paulogdm) [reply packages/now-cli/src/types.ts:204]: Only now I saw this comment after fixing it. It took me a few minutes...
Thanks @styfle !

**Review by styfle — DISMISSED** [review]
Thanks!

**Review by styfle — COMMENTED** [review]
@paulogdm Can you add an integration test for this new behavior?

**Comment by paulogdm** [comment]
@styfle Sure. Give me the rest of the week to do so.

**Comment by paulogdm** [comment]
Added tests and also tested this locally...
I could use some advice on the best patterns for those tests in this repository.

**Review by styfle — APPROVED** [review]
Amazing, thanks! 🎉

File: packages/now-cli/src/commands/dns/add.ts [hunk 11]
```diff
@@ -54,9 +55,34 @@ export default async function add(
...
+          chalk.bold(`${argData.type} ${argData.srv.priority} ${argData.srv.port} ${argData.srv.weight} ${argData.srv.target}`)
+        );
+        break;
+      default:
+        output.log(
+          chalk.bold(`${argData.type} ${argData.value}`)
+        );
+        break;
+    }
+    if (!await promptBool(`Do you want to create this record on "${chalk.bold(argData.name)}.${chalk.bold(domain)}?"`)) {
```
Comment (AndyBitz) [inline packages/now-cli/src/commands/dns/add.ts:79]:
```suggestion
    if (!(await promptBool(`Do you want to create this record on "${chalk.bold(argData.name)}.${chalk.bold(domain)}?"`))) {
```

Comment (AndyBitz) [reply packages/now-cli/src/commands/dns/add.ts:79]: Do we also have a `skip` flag like `-y` for this to make sure it works without user interaction?

File: packages/now-cli/src/util/dns/get-dns-data.ts [hunk 12]
```diff
@@ -1,10 +1,31 @@
 import chalk from 'chalk';
-import { DNSRecordData } from '../../types';
+import {
+  DNSRecordData,
+  ARecordData,
+  AAAARecordData,
+  ALIASRecordData,
+  CAARecordData,
+  CNAMERecordData,
+  TXTRecordData,
```
Comment (AndyBitz) [inline packages/now-cli/src/util/dns/get-dns-data.ts:9]: Looks like the imports from line 4 to line 9 are not used

File: packages/now-cli/src/util/dns/get-dns-data.ts [hunk 13]
```diff
@@ -16,82 +37,95 @@ export default async function getDNSData(
...
-          value
-        }
-      : null;
-  } catch (error) {
+  const value = await getTrimmedString(`- ${type} value: `);
+  output.log(`${chalk.cyan(name)} ${chalk.bold(type)} ${chalk.cyan(value)}`);
+  if (!(await verifyData())) {
     return null;
   }
+  const record = {
```
Comment (AndyBitz) [inline packages/now-cli/src/util/dns/get-dns-data.ts:114]: Do we have a type for this? Otherwise we could just use `return { ... }` here

File: packages/now-cli/src/commands/dns/add.ts (hunk 11 above)
Comment (paulogdm) [reply packages/now-cli/src/commands/dns/add.ts:79]: Nice catch! I will add that...
